kimchi/src/circuits/polynomials/generic.rs

```diff
@@ -1,73 +1,284 @@
+//! |:--:|:--:|:--:|:--:|:--:|:--:|:-:|:-:|:-:|:-:|:--:|:--:|:--:|:--:|:--:|:--:|
+//! | l1 | r1 | o1 | m1 | c1 | l2 | r2 | o2 | m2 | c2  |    |    |    |    |    |
+//!
+//! with m1 (resp. m2) the mul selector for the first (resp. second) gate,
+//! and c1 (resp. c2) the constant selector for the first (resp. second) gate.
+//!
+//! The polynomial looks like this:
+//!
+//! <pre>
+//! [
+//!   alpha1 * (w0 * coeff0 + w1 * coeff1 + w2 * coeff2 + w0 * w1 * coeff3 + coeff4) +
+//!   alpha2 * (w3 * coeff5 + w4 * coeff6 + w5 * coeff7 + w3 w4 coeff8 + coeff9)
+//! ] * generic_selector
+//! </pre>
+//!
 
 use crate::circuits::{
     constraints::ConstraintSystem,
-    gates::generic::{CONSTANT_COEFF, MUL_COEFF},
+    gate::{CircuitGate, GateType},
     polynomial::COLUMNS,
-    wires::GENERICS,
+    wires::{GateWires, GENERICS},
 };
 use ark_ff::{FftField, SquareRootField, Zero};
-use ark_poly::{
-    univariate::DensePolynomial, EvaluationDomain, Evaluations, Polynomial,
-    Radix2EvaluationDomain as D,
-};
+use ark_poly::{univariate::DensePolynomial, Evaluations, Radix2EvaluationDomain as D};
 use array_init::array_init;
 use rayon::prelude::*;
 
 /// Number of constraints produced by the gate.
-pub const CONSTRAINTS: usize = 1;
+pub const CONSTRAINTS: usize = 2;
+
+/// Offset for the second generic gate
+pub const GENERICS_COEFFS: usize = GENERICS + 1 /* mul */ + 1 /* cst */;
+
+/// The different type of computation that are possible with a generic gate.
+/// This type is useful to create a generic gate via the [create_generic_easy] function.
+pub enum GenericGate<F> {
+    /// Add two values.
+    Add {
+        /// Optional coefficient that can be multiplied with the left operand.
+        left_coeff: Option<F>,
+        /// Optional coefficient that can be multiplied with the right operand.
+        right_coeff: Option<F>,
+        /// Optional coefficient that can be multiplied with the output.
+        output_coeff: Option<F>,
+    },
+    /// Multiplication of two values
+    Mul {
+        /// Optional coefficient that can be multiplied with the output.
+        output_coeff: Option<F>,
+        /// Optional coefficient that can be multiplied with the multiplication result.
+        mul_coeff: Option<F>,
+    },
+    /// A constant, the constructor contains the constant itself
+    Const(F),
+    /// A public gate
+    Pub,
+}
+
+impl<F: FftField> CircuitGate<F> {
+    /// This allows you to create two generic gates that will fit in one row, check [create_generic_easy] for a better to way to create these gates.
+    pub fn create_generic(wires: GateWires, c: [F; GENERICS * 2 + 2 + 2]) -> Self {
+        CircuitGate {
+            typ: GateType::Generic,
+            wires,
+            coeffs: c.to_vec(),
+        }
+    }
+
+    /// This allows you to create two generic gates by passing the desired
+    /// `gate1` and `gate2` as two [GenericGate].
+    pub fn create_generic_easy(
```

Comment:
or `create_generic_gadget`, if we start using that term everywhere

Reply:
I'm not sure I like this terminology, since you're 'using' the gadget here rather than 'creating' it. (My mental model is that the gadget is the code that generates the constraint(s), not the generated constraints themselves.)

Reply:
In my mind a gadget is a series of rows that you can abstract and re-use without knowing what's in the rows. I'm not sure it makes the most sense for the generic gate since it's a single row, but for poseidon the gadget is the 11 rows that you get (I called that gadget there as well https://github.com/o1-labs/proof-systems/blob/master/kimchi/src/circuits/gates/poseidon.rs#L67). So the function creates the gadget.

I might have taken that definition from there: https://zcash.github.io/halo2/concepts/gadgets.html